File: haskell/2618.hs
import Control.Monad (replicateM)
import Control.Monad.ST ( ST, runST )
import Data.Array.ST (STArray, MArray (newArray), readArray, writeArray)
import Data.Array (Array, listArray, (!))

data Coord = Coord {
  r :: Int,
  c :: Int
}

data Solver s = Solver {
  n :: Int,
  w :: Int,
  dp :: STArray s (Int, Int) Int,
  locs :: Array Int Coord
}

l1Dist :: Coord -> Coord -> Int
l1Dist c1 c2 = abs (r c1 - r c2) + abs (c c1 - c c2)

newSolver :: [Coord] -> Int -> Int -> ST s (Solver s)
newSolver locs n w = do
  dpArr <- newArray ((0, 0), (w, w)) (-1)
  let locsArr = listArray (0, w) (Coord 0 0 : locs)
  return $ Solver n w dpArr locsArr

findMinDist :: Solver s -> Int -> Int -> ST s Int
findMinDist solver p1 p2
  | p1 == w solver || p2 == w solver = return 0
  | otherwise = do
      minDist <- readArray (dp solver) (p1, p2)
      if minDist /= -1
        then return minDist
        else do
          let cur  = max p1 p2
              next = cur + 1

          dist1 <- if p1 == 0
                   then return $ l1Dist (Coord 1 1) (locs solver ! next)
                   else return $ l1Dist (locs solver ! p1) (locs solver ! next)
          rest1 <- findMinDist solver next p2

          dist2 <- if p2 == 0
                   then return $ l1Dist (Coord (n solver) (n solver)) (locs solver ! next)
                   else return $ l1Dist (locs solver ! p2) (locs solver ! next)
          rest2 <- findMinDist solver p1 next

          let res = min (dist1 + rest1) (dist2 + rest2)
          writeArray (dp solver) (p1, p2) res
          return res

findTrace :: Solver s -> Int -> Int -> ST s [Int]
findTrace solver p1 p2
  | p1 == w solver || p2 == w solver = return []
  | otherwise = do
      let cur  = max p1 p2
          next = cur + 1

      dist1 <- if p1 == 0
               then return $ l1Dist (Coord 1 1) (locs solver ! next)
               else return $ l1Dist (locs solver ! p1) (locs solver ! next)
      rest1 <- readArray (dp solver) (next, p2)

      dist2 <- if p2 == 0
               then return $ l1Dist (Coord (n solver) (n solver)) (locs solver ! next)
               else return $ l1Dist (locs solver ! p2) (locs solver ! next)
      rest2 <- readArray (dp solver) (p1, next)

      let p1Dist = dist1 + rest1
          p2Dist = dist2 + rest2

      remTrace <- if p1Dist < p2Dist
                  then findTrace solver next p2
                  else findTrace solver p1 next
      
      return $ (if p1Dist < p2Dist then 1 else 2) : remTrace

main :: IO ()
main = do
  n <- readLn :: IO Int
  w <- readLn :: IO Int

  locs <- replicateM w $ do
    [r, c] <- fmap (map read . words) getLine :: IO [Int]
    return $ Coord r c
  
  let (minDist, route) = runST $ do
        solver <- newSolver locs n w
        minDist <- findMinDist solver 0 0
        route <- findTrace solver 0 0
        return (minDist, route)
  
  print minDist
  mapM_ print route

File: rust/src/bin/2618.rs
use io::Write;
use std::{io, str, cmp};

struct UnsafeScanner<R> {
    reader: R,
    buf_str: Vec<u8>,
    buf_iter: str::SplitAsciiWhitespace<'static>,
}

impl <R: io::BufRead> UnsafeScanner<R> {
    fn new(reader: R) -> Self {
        Self {
            reader,
            buf_str: vec![],
            buf_iter: "".split_ascii_whitespace(),
        }
    }

    fn token<T: str::FromStr>(&mut self) -> T {
        loop {
            if let Some(token) = self.buf_iter.next() {
                return token.parse().ok().expect("Failed parse");
            }
            self.buf_str.clear();
            self.reader
                .read_until(b'\n', &mut self.buf_str)
                .expect("Failed read");
            self.buf_iter = unsafe {
                let slice = str::from_utf8_unchecked(&self.buf_str);
                std::mem::transmute(slice.split_ascii_whitespace())
            }
        }
    }
}

struct Coord {
    r: i32,
    c: i32,
}

#[inline(always)]
fn l1_dist(c1: &Coord, c2: &Coord) -> i32 {
    (c1.r - c2.r).abs() + (c1.c - c2.c).abs()
}

struct Solver<'a> {
    n: i32,
    w: usize,
    route: Vec<i32>,
    dp: Vec<Vec<i32>>,
    locs: &'a Vec<Coord>,
}

impl<'a> Solver<'a> {
    fn new(locs: &'a Vec<Coord>, n: i32, w: usize) -> Self {
        let dp = vec![vec![-1; w + 1]; w + 1];
        let route = vec![0; w + 1];
        Self { n, w, route, dp, locs, }
    }

    fn find_min_travel_dist(
        &mut self,
        police1: usize,
        police2: usize,
    ) -> i32 {
        if police1 == self.w || police2 == self.w {
            return 0;
        }

        if self.dp[police1][police2] != -1 {
            return self.dp[police1][police2];
        }

        let cur_inc = cmp::max(police1, police2);
        let next_inc = cur_inc + 1;

        // The case police1 is inspecting the next incident.
        let mut dist1;
        if police1 == 0 {
            dist1 = l1_dist(&Coord { r: 1, c: 1 }, &self.locs[next_inc]);
        } else {
            dist1 = l1_dist(&self.locs[police1], &self.locs[next_inc]);
        }
        dist1 += self.find_min_travel_dist(next_inc, police2);

        // The case police2 is inspecting the next incident.
        let mut dist2;
        if police2 == 0 {
            dist2 = l1_dist(&Coord { r: self.n, c: self.n }, &self.locs[next_inc]);
        } else {
            dist2 = l1_dist(&self.locs[police2], &self.locs[next_inc]);
        }
        dist2 += self.find_min_travel_dist(police1, next_inc);

        self.dp[police1][police2] = cmp::min(dist1, dist2);

        self.dp[police1][police2]
    }

    fn find_trace(
        &mut self,
        police1: usize,
        police2: usize,
    ) {
        if police1 == self.w || police2 == self.w {
            return;
        }
        
        let cur_inc = cmp::max(police1, police2);
        let next_inc = cur_inc + 1;

        // The case police1 is inspecting the next incident.
        let mut dist1;
        if police1 == 0 {
            dist1 = l1_dist(&Coord { r: 1, c: 1 }, &self.locs[next_inc]);
        } else {
            dist1 = l1_dist(&self.locs[police1], &self.locs[next_inc]);
        }
        dist1 += self.dp[next_inc][police2];

        // The case police2 is inspecting the next incident.
        let mut dist2;
        if police2 == 0 {
            dist2 = l1_dist(&Coord { r: self.n, c: self.n }, &self.locs[next_inc]);
        } else {
            dist2 = l1_dist(&self.locs[police2], &self.locs[next_inc]);
        }
        dist2 += self.dp[police1][next_inc];

        if dist1 < dist2 {
            self.route[cur_inc] = 1;
            self.find_trace(next_inc, police2);
        } else {
            self.route[cur_inc] = 2;
            self.find_trace(police1, next_inc);
        }
    }
}

fn main() {
    let (stdin, stdout) = (io::stdin(), io::stdout());
    let mut scan = UnsafeScanner::new(stdin.lock());
    let mut out = io::BufWriter::new(stdout.lock());

    let (n, w) = (scan.token::<i32>(), scan.token::<usize>());

    let mut locs: Vec<Coord> = vec![];
    locs.push(Coord { r: 0, c: 0 });
    for _ in 1..=w {
        locs.push(Coord {
            r: scan.token::<i32>(),
            c: scan.token::<i32>(),
        });
    }

    let mut solver = Solver::new(&locs, n, w);
    let min_travel_dist = solver.find_min_travel_dist(0, 0);
    writeln!(out, "{min_travel_dist}").unwrap();

    solver.find_trace(0, 0);
    for i in 0..w {
        writeln!(out, "{}", solver.route[i]).unwrap();
    }
}